scipy/optimize/minpack.py

```diff
@@ -663,10 +675,26 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
         else:
             xdata = np.asarray(xdata)
 
-    weights = 1.0 / asarray(sigma) if sigma is not None else None
-    func = _wrap_func(f, xdata, ydata, weights)
+    # Determine type of sigma
+    if sigma is not None:
+        sigma = np.asarray(sigma)
+
+        if sigma.shape == (ydata.size, ydata.size):
+            from numpy.linalg import LinAlgError
```

Comment:
I was trying to follow the same style as used in the leastsq function where LinAlgError was imported where it was used. I can fix that everywhere if you think it is appropriate.